### PR TITLE
feat(vscode): show docs on hover

### LINKED ISSUE #1569: VSCode hover for methods and closures should show parameter names
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### Feature Spec

When hovering over a symbol representing a function of some kind (method or closure), it should show the names of parameters rather than just the types.

### Use Cases

Ugliness

### Implementation Notes

_No response_

### Component

IDE Extension

### PR DESCRIPTION
**Pull request by eladb**
## Description

Add initial support for hover documentation for classes, functions and symbols.

Screenshot 2023-06-09 at 16 24 17

Screenshot 2023-06-09 at 16 24 35

Screenshot 2023-06-09 at 16 25 34

## Implementation Notes

Propagate the JSII docs throughout the type system. This is currently done for function signatures and parameters as well as classes.

Created a `Documented` trait and implement it for various types in order to allow polymorphic doc generation.
Removed all the spans from the synthetic syntax produced by the `ClosureTransformer` since they shadowed the actual symbols in the source file and that caused the hover visitor to ignore any statements within the transformed closure.

Rewrite HoverVisitor to find the relevant element and use the new trait to render docs.

Added `name` to `FunctionParameter` so it will be possible to display the parameter names in docs.

Added the desugaring phase to the LSP so that closure will be transformed appropriately.

## Misc

* Removed special handling of the `core.Inflight` type which no longer exists.
* Add `full_path_str` to `UserDefinedType` to make it easier to produce a nested string for lookups.
* Extended visitor to also allow subscribing to symbol visits (low level).
* Removed the name of the interface when showing closure signatures. This fixes https://github.com/winglang/wing/issues/2786.
* Do not apply spans to default initializers of classes.
* Fix #1569 by adding parameter names to function signatures.

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

**Comment by eladb** [comment]
@MarkMcCulloh any chance you can take a look at this? I have a couple of tests that are crashing with some memory issue (see "TODO" above).

**Comment by MarkMcCulloh** [comment]
@eladb  Added several changes to the PR to address the test failures, as well as a couple other bugs:

- resolve latest merge conflict
- `invalid memory` was caused by the usage of `deref()` rather than (I assume) `.as_ref()`
- `inside_class_field` now shows info, I added back in the code that was there before to handle scope tracking in classes
- Added additional special hover cases (for loops, try/catch, etc.) to make sure lookups succeed
- When compiling in the lsp, the `scope` needs to be boxed prior to type checking. Without this, lookups would fail when vars were defined in the parent due to pointer issues.
- In fold_scope, scope span was being cleared but it's needed for most lsp functions
- In completions, the highlights for which argument is current was broken due to the docs
- For primitive types, the docs should be empty until we figure that out. Otherwise the redundancy was strange (it was just printing the type name twice)

**Comment by eladb** [comment]
@MarkMcCulloh 🙏🙏🙏

File: libs/wingc/src/lsp/hover.rs [hunk 1]
```diff
@@ -25,26 +28,24 @@ impl<'a> HoverVisitor<'a> {
...
 		self.visit_scope(self.current_scope);
 		self.found.clone()
 	}
 
 	/// Try to look up a full path of a symbol in the current scope and if found, render the docs
 	/// associated with the symbol kind. Returns `None` if not found.
 	fn lookup_docs(&mut self, nested_str: &str, property: Option<&Symbol>) -> Option<String> {
-		let b = self.current_scope.env.borrow();
+		let current_env = self.current_scope.env.borrow();
+		let current_env = current_env.as_ref()?;
```
Comment (eladb) [inline libs/wingc/src/lsp/hover.rs:46]: TIL @MarkMcCulloh thanks!

File: libs/wingc/src/lsp/snapshots/hovers/static_method.snap [hunk 2]
```diff
@@ -0,0 +1,14 @@
+---
+source: libs/wingc/src/lsp/hover.rs
+---
+contents:
+  kind: markdown
+  value: "```wing\nmy: preflight (): str\n```\n---"
```
Comment (eladb) [inline libs/wingc/src/lsp/snapshots/hovers/static_method.snap:6]: missing `static` here?

File: libs/wingc/src/lsp/hover.rs [hunk 3]
```diff
@@ -106,40 +149,132 @@ impl<'a> Visit<'a> for HoverVisitor<'a> {
...
+				if sym.span.contains(&self.position) {
+					self.found = Some((sym.span.clone(), self.lookup_docs(&sym.name, None)));
+				}
+			}
+			Reference::InstanceMember {
+				object,
+				property,
+				optional_accessor: _,
+			} => {
+				if let Some(obj_type) = self.types.get_expr_type(object) {
```
Comment (eladb) [inline libs/wingc/src/lsp/hover.rs:246]: what about the case where `object` is matched?

File: libs/wingc/src/lsp/snapshots/hovers/static_method.snap (hunk 2 above)
Comment (eladb) [reply libs/wingc/src/lsp/snapshots/hovers/static_method.snap]: added

File: libs/wingc/src/lsp/hover.rs (hunk 3 above)
Comment (eladb) [reply libs/wingc/src/lsp/hover.rs:244]: checked

**Comment by eladb** [comment]
@MarkMcCulloh ready for review

File: libs/wingc/src/parser.rs [hunk 4]
```diff
@@ -890,14 +889,22 @@ impl<'s> Parser<'s> {
...
-					.named_children(&mut cursor)
-					.filter_map(|param_type| self.build_type_annotation(&param_type, phase).ok())
-					.collect::<Vec<TypeAnnotation>>();
+
+				let mut parameters = vec![];
+				for param_type in param_type_list_node.named_children(&mut cursor) {
+					let t = self.build_type_annotation(&param_type, phase)?;
+
+					parameters.push(FunctionParameter {
+						name: "".into(),
```
Comment (Chriscbr) [inline libs/wingc/src/parser.rs:898]: Usually when a value isn't always available, it's preferable to model it in Rust using the `Option` type, e.g. `Option<String>`

File: libs/wingc/src/ast.rs [hunk 5]
```diff
@@ -140,19 +141,10 @@ pub enum TypeAnnotationKind {
 	MutMap(Box<TypeAnnotation>),
 	Set(Box<TypeAnnotation>),
 	MutSet(Box<TypeAnnotation>),
-	Function(FunctionTypeAnnotation),
```
Comment (Chriscbr) [inline libs/wingc/src/ast.rs:143]: What does the VS code extension show you if you hover over a variable/field with a function type, like "fn" in the following example?

```js
class A {
  fn: inflight (num): str;
}
```

**Review by MarkMcCulloh — APPROVED** [review]
LGTM, really excited for this change.

I pushed up a fix for completions inside inflights/test statements (they were incorrect due to the span changes in the fold thing). 
Also added a small misc change to filter out some symbols from completions.

Added a do-not-merge label to give you a chance to counter-review my changes to your branch.

File: libs/wingc/src/ast.rs (hunk 5 above)
Comment (eladb) [reply libs/wingc/src/ast.rs:143]: It shows `p1` as the name of the argument. Not ideal, but good enough I think.

**Comment by eladb** [comment]
> 

Looks good!

**Pull request merged**